From 854da91ca9d491e553e72067f7116577199cfce8 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Sun, 20 Jun 2021 23:36:09 +0300
Subject: [PATCH] [SW-46313] gaudi: security: make sf memmap dyn cfg

Add support for enabling the read-only access to the memory-mapped SPI
flash reagion.
The preboot_security_sf_pci_set_ro() function can be used to
enable/disable the R/O access.

Change-Id: I11c7ab805894355f0d1aec7da2af6ed42e10ae4d
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/arch/gaudi/security.c b/arch/gaudi/security.c
index eb1929e..5c98e17 100644
--- a/arch/gaudi/security.c
+++ b/arch/gaudi/security.c
@@ -1659,6 +1659,7 @@
 	VERS_END_TO_IMG,
 	IMAGE,
 	IMG_END_TO_SRAM_END,
+	STM_SF_MEMMAP,
 	SPSRAM_IMAGE,
 };
 
@@ -1693,6 +1694,11 @@
 		.rr = PCIE_DYN_RR_SRAM_END,
 		.min = 0,
 		.max = SRAM_END,
+	},
+	[STM_SF_MEMMAP] = {
+		.rr = PCIE_DYN_RR_STM_SF,
+		.min = mmSTM_BASE - 1,
+		.max = HL_SPI_END,
 	},
 	[SPSRAM_IMAGE] = {
 		.rr = PCIE_DYN_RR_SPSRAM_IMG,
@@ -1964,3 +1970,16 @@
 				  HBM_RR_MASK, true);
 	preboot_hbw_rr_set_rtr_ar(HBW_RR_HBM, PCI_RTR_NO, 0, 0, false);
 }
+
+void preboot_security_sf_pci_set_ro(bool en)
+{
+	struct preboot_sec_dyn_pcie_rr_cfg *cfg;
+
+	if (!preboot_security_is_enabled())
+		return;
+
+	cfg = &preboot_sec_dyn_cfg[STM_SF_MEMMAP];
+	preboot_secure_pcie_rr_wr(cfg->rr, cfg->min, cfg->max);
+	preboot_secure_pcie_rr_rd(cfg->rr, cfg->min,
+				  en ? HL_SPI_BASE : cfg->max);
+}
diff --git a/include/security.h b/include/security.h
index bbca703..61e2cfb 100644
--- a/include/security.h
+++ b/include/security.h
@@ -21,5 +21,6 @@
 void preboot_security_cleanup_before_run(void);
 void preboot_security_sp_sram_write_key_hash(u8 *key, size_t len);
 void preboot_security_hbm_pci_open_ro(void);
+void preboot_security_sf_pci_set_ro(bool en);
 
 #endif /* _HL_SECURITY_H_ */
